torch/masked/maskedtensor/functions.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import math
+
+import torch
+from torch.nn.functional import dropout, linear
+
+from .matmul import masked_bmm
+
+__all__ = ['multi_head_attention_forward']
+
+def _in_projection_packed(q, k, v, w, b):
+    w_q, w_k, w_v = w.chunk(3)
+    if b is not None:
+        raise ValueError("b must be None")
+    b_q = b_k = b_v = None
+    return linear(q, w_q, b_q), linear(k, w_k, b_k), linear(v, w_v, b_v)
+
+
+def _scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=0.0):
+    B, Nt, E = q.shape
+    q = q / math.sqrt(E)
+    # (B, Nt, E) x (B, E, Ns) -> (B, Nt, Ns)
+    attn = masked_bmm(q, k.transpose(-2, -1), attn_mask)
+    attn = torch.nn.functional.softmax(attn, dim=-1)
+    if dropout_p > 0.0:
+        attn = dropout(attn, p=dropout_p)
+    # (B, Nt, Ns) x (B, Ns, E) -> (B, Nt, E)
+    output = torch.bmm(attn, v)
+    return output, attn
+
+
+def multi_head_attention_forward(
```

Comment:
are we testing this? the tests for the ordinary version are pretty extensive. (I don't imagine it's just meant to be an example of how to use MTs in MHA, since it's basically a tweaked version of the full version in nn.functional)

Reply:
This is implicitly tested in the `test_mha_41508` and the code is very similar to the `torch.nn.functional` code

Reply:
sorry, what's `test_mha_41508` 😬